lib/services/database/challenge_repository_service.dart

```diff
@@ -0,0 +1,175 @@
+  static const double maxChallengeRadius = 3; // in km
+  static const double minChallengeRadius = 0.5;
+  static const maxChallengeDuration = Duration(days: 1);
+
+  /// Creates a new challenge repository service
+  /// with the given [firestore] and [postRepositoryService]
+  ChallengeRepositoryService({
+    required FirebaseFirestore firestore,
+    required PostRepositoryService postRepositoryService,
+  })  : _firestore = firestore,
+        _postRepositoryService = postRepositoryService;
+
+  /// Completes the challenge of the user with id [uid] on the post with id [pid]
+  Future<void> completeChallenge(
+    UserIdFirestore uid,
+    PostIdFirestore pid,
+  ) async {
+    final userDocRef =
+        _firestore.collection(UserFirestore.collectionName).doc(uid.value);
+    await _completeChallenge(userDocRef, pid);
+  }
+
+  Future<void> _completeChallenge(
+    DocumentReference parentRef,
+    PostIdFirestore pid,
+  ) async {
+    parentRef
+        .collection(ChallengeFirestore.subCollectionName)
+        .doc(pid.value)
+        .update({
+      ChallengeData.isCompletedField: true,
+    });
+  }
+
+  /// Returns the active challenges of the user with id [uid] that is located at [pos]
+  Future<List<ChallengeFirestore>> getChallenges(
+    UserIdFirestore uid,
+    GeoPoint pos,
+  ) async {
+    final userDocRef =
+        _firestore.collection(UserFirestore.collectionName).doc(uid.value);
+
+    return _getChallenges(userDocRef, pos);
+  }
+
+  Future<List<ChallengeFirestore>> _getChallenges(
+    DocumentReference parentRef,
+    GeoPoint pos,
+  ) async {
+    final challengesCollectionRef =
+        parentRef.collection(ChallengeFirestore.subCollectionName);
+    final pastChallengesCollectionRef = parentRef
+        .collection(ChallengeFirestore.pastChallengesSubCollectionName);
+
+    final challengesSnap = await challengesCollectionRef.get();
+    final pastChallengesSnap = await pastChallengesCollectionRef.get();
+
+    final pastPostIds = pastChallengesSnap.docs
+        .map((doc) => PostIdFirestore(value: doc.id))
+        .toSet();
+
+    final List<ChallengeFirestore> activeChallenges =
+        List.empty(growable: true);
+
+    for (final challengeSnap in challengesSnap.docs) {
+      final challenge = ChallengeFirestore.fromDb(challengeSnap);
+      if (challenge.data.isExpired ||
+          !await _postRepositoryService.postExists(challenge.postId)) {
+        _deleteChallenge(challengeSnap, parentRef);
+        pastPostIds.add(challenge.postId);
+      } else {
+        activeChallenges.add(challenge);
+      }
+    }
+
+    final now = DateTime.now();
+    final sum = now.add(maxChallengeDuration);
+    final expiresOn = DateTime(
+      sum.year,
+      sum.month,
+      sum.day,
+    ); // truncates to the day
+
+    final possiblePosts = await inRangeUnsortedPosts(pos);
+    final postIt = possiblePosts.iterator;
+    final activePostIds = activeChallenges.map((e) => e.postId).toSet();
```

Comment:
Like @yoannLafore said, we could add a little bit of clarity also here.
```suggestion
    final activePostIds = activeChallenges.map((challenge) => challenge.postId).toSet();
```